Assignments/ip1.md

```diff
@@ -0,0 +1,301 @@
+<details><summary markdown="span">View the specification for these methods</summary>
+{% highlight typescript %}
+  /**
+   * Adds a player to the game.
+   * Updates the game's state to reflect the new player.
+   * If the game is now full (has two players), updates the game's state to set the status to IN_PROGRESS.
+   *
+   * @param player The player to join the game
+   * @throws InvalidParametersError if the player is already in the game (PLAYER_ALREADY_IN_GAME_MESSAGE)
+   *  or the game is full (GAME_FULL_MESSAGE)
+   */
+  public _join(player: Player): void
+
+   /**
+   * Removes a player from the game.
+   * Updates the game's state to reflect the player leaving.
+   * If the game has already started, updates the game's status to OVER and sets the winner to the other player.
+   * If the game has not started yet, updates the game's status to WAITING_TO_START.
+   *
+   * @param player The player to remove from the game
+   * @throws InvalidParametersError if the player is not in the game (PLAYER_NOT_IN_GAME_MESSAGE)
+   */
+  protected _leave(player: Player): void
+{% endhighlight %}
+</details>
+
+*Testing*: Avery has provided you with some very simple (and incomplete) tests for `_join` and `_leave`. You can run these tests by running the command `npx jest --watch TicTacToeGame.test`, which will automatically re-run the tests as you update the file (note that tests for `applyMove` will also run - but you can ignore those at this point). You should improve the `_join` and `_leave` tests to check the entire specification. Please implement these additional tests in the file `src/town/games/TicTacToeGame.test.ts`.
+
+Grading for implementation tasks:
+* `_join`: 6 points
+* `_leave`: 6 points
+
+Grading for testing tasks:
+* `_join`:
+  * 6 points for detecting all 25 faults, or
+  * 3 points for detecting 24 faults
+* `_leave`: 6 points 
+
+### Task 2: Tic-Tac-Toe Game Semantics (50 points total)
+The next (and largest) task for this deliverable is to implement the method `TicTacToeGame.applyMove`, which applies a player's move to the game. This method is responsible for validating the move, and updating the game state to reflect the move. Given the complexity of this method, you should anticipate creating (at least one) private, helper method to implement its logic.
+
+<details><summary markdown="span">View the specification for this method</summary>
+{% highlight typescript %}
+  /* 
+   * Applies a player's move to the game.
+   * Uses the player's ID to determine which game piece they are using (ignores move.gamePiece)
+   * Validates the move before applying it. If the move is invalid, throws an InvalidParametersError with
+   * the error message specified below.
+   * A move is invalid if:
+   *    - The move is on a space that is already occupied (use BOARD_POSITION_NOT_EMPTY_MESSAGE)
+   *    - The move is not the player's turn (MOVE_NOT_YOUR_TURN_MESSAGE)
+   *    - The game is not in progress (GAME_NOT_IN_PROGRESS_MESSAGE)
```

Comment:
A move should also be invalid if the `move.gamePiece` field of the move itself is not compatible with the current player's game piece, because when checking for a winner later there's no other way to pull that information from the `moves` array except by checking whether a particular move's position in the array has even or odd parity, which I doubt was the intention here.

Reply:
Perhaps this is a surprise also, but earlier in the doc for applyMove it says that you should ignore the gamePiece specified on the move and determine the game piece by comparing the player to the game's state. Hence, I don't think that there is any checking needed to see if the gamePiece is compatible, right?

Reply:
> earlier in the doc for applyMove it says that you should ignore the gamePiece specified on the move and determine the game piece by comparing the player to the game's state

I did see this. I think the reason for my comment on this was to highlight the fact that this requirement makes two parts of the implementation significantly harder, and double-check that that is intentional:
* determining the game piece for the *current* move by comparing the player to the game's state
* determining the game piece for *all past moves* by checking the parity of the moves stored in the game's state itself

The first part is clearly intentional. It wasn't clear to me that the second part is intentional (but I'm guessing that it is, based on your comment, in which case we can ignore this).